[FIX] 926 auto ensure all models support alias and 924 Configuring hyperparameter space for Auto* Models

elephaint opened the pull request:
Fixes #926 and fixes #924.

- Fix of #926:
Fixed Auto* models that do not support all required variables from BaseAuto
Add unit test to all Auto* models to test if the Auto* models contain all required variables from BaseAuto
Some reformatting of models.ipynb, putting unit tests per model in hidden cell below each model
- Fix of #924:
Added a `get_default_config` method for all Auto* models
Fixed incompatible hyperparameter configurations in default_config and in three base models
Added unit tests to all auto models to ensure they all run with default configs, on both backends.
Update of Hyperparameter tuning guide to demonstrate the use of `get_default_config`

jmoralez commented on nbs/models.ipynb:
```diff
@@ -83,6 +83,7 @@
     "\n",
```
<pre class="ql-syntax" spellcheck="false">Line #48.            reconciliation = config.pop('reconciliation')
</pre><p>It isn't possible to do this check with optuna, so this model doesn't suport it. Can you please add a check for the backend in the init and if it's not ray raise an error? Or if you can make it work that'd be awesome, I just couldn't think of a way to remove this key from the function body.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Nixtla/neuralforecast/pull/927/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='97'/>

elephaint replied:
I think this boils down to essentially the same issue as your other remark: Optuna just doesn't work with Multivariate models and AutoHINT, as we need to enforce some parameters in the config. Let me think what would be best to solve it.

elephaint replied:
I included an exception for now, I haven't found a way to get Optuna to work, it fails in base_auto on `metrics = trainer.callback_metrics`, which is `None` when using the redefined `_fit_model` of `AutoHINT` (and haven't figured out how to correct this.)

jmoralez replied:
Oh that should be easy to fix, that's the second argument of the return from _fit_model in AutoHINT, so you should go to HINT and store the trainer that is returned from the self.model.fit and propagate that back. I know it's hideous haha but I'm already working on improving it.

jmoralez commented on nbs/models.ipynb:
```diff
@@ -83,6 +83,7 @@
     "\n",
```
<pre class="ql-syntax" spellcheck="false">Line #47.                    config['n_series'] = n_series
</pre><p>I think this would work as well if you set it before the <code>if backend == 'optuna':</code> and then we can get rid of the <code>if backend == 'ray':</code>below. Same for the other ones</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Nixtla/neuralforecast/pull/927/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='102'/>

jmoralez replied:
Ah I see, I thought we were only setting it for the default config. We should probably remove it from the auto model then so that the user always provides it through the config (and an error will be raised if it's not provided). WDYT?

elephaint replied:
I came up with a method of checking the supplied config for `'n_series' = n_series` in case of Optuna backend, by using the MockTrial you created earlier. It raises an Exception if the user does not supply  `'n_series' = n_series`.

elephaint replied:
Sorry, I missed your last comment (should have renewed the page). Yes, I can change it so that it always throws an Exception (now it only throws an Exception when using Optuna).

wdyt?

jmoralez replied:
Removing it from the init signature is a breaking change, but I think that's the better option for the future. We can deprecate it and issue a FutureWarning indicating that the n_series should be provided through the config if it's provided in the init signature for the multivariate auto models.

elephaint replied:
@jmoralez In general, why do we ever want the user to supply `n_series` in MultiVariate models? We can infer it ourselves from `unique_id` in the dataset, right? In what situation is it not true that `n_series == unique(unique_id)`?

(This is more in the sense that if we make a breaking change to remove it, why not remove it entirely across all MultiVariate models)

elephaint replied:
Fiddled around with it a bit but think it's indeed better to remove `n_series ` everywhere (i.e., also from the base multivariate models) rather than only from Auto* models, if we want to make that breaking change. 

wdyt?

elephaint replied:
Ok, then I suggest we leave it as it is now (also for Auto*)

jmoralez replied:
Agree

jmoralez reviewed (approved):
LGTM! Let's just please also wait for @cchallu's review as well.

elephaint commented:
> LGTM! Let's just please also wait for @cchallu's review as well.

Thanks - it does not fix the default_config issue yet, so I'm changing a few things (converted to draft)

elephaint commented:
- Added a `get_default_config` method for all Auto* models
- Fixed incompatible hyperparameter configurations in base models
- Added unit tests to all auto models to ensure they all run with default configs, on both backends.

jmoralez commented on nbs/examples/Automatic_Hyperparameter_Tuning.ipynb:
```diff
@@ -181,11 +181,23 @@
    "cell_type": "markdown",
```
<pre class="ql-syntax" spellcheck="false">Line #4.        config = {**nhits_default_config(trial)}
</pre><p>The function already returns a dict, so there's no need to unpack it, you can assign it to the config variable directly or unpack it and return it in a single statement like:</p><pre class="ql-syntax" spellcheck="false">def config_nhits(trial):
    return {
        **nhits_default_config(trial),
        'random_seed': trial_suggest_int(...),
        ...
    }
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Nixtla/neuralforecast/pull/927/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

elephaint replied:
Thanks - agree although that solution implicitly requires users to perform the operation in exactly that order, i.e.

```
return {
    **nhits_default_config(trial),
    'random_seed': trial_suggest_int(...),
    ...
}
```
works, but this:

```
return {
    'random_seed': trial_suggest_int(...),
    **nhits_default_config(trial),
    ...
}
```
will not correctly update the parameters. That's why I thought to make the order and update step explicit.

jmoralez reviewed (approved):
Just an optional comment
